flink-ml-lib/src/main/java/org/apache/flink/ml/classification/naivebayes/NaiveBayesModel.java

```diff
@@ -0,0 +1,223 @@
+                    return stream.transform(
+                            this.getClass().getSimpleName(),
+                            outputTypeInfo,
+                            new PredictLabelOperator(
+                                    new PredictLabelFunction(featuresCol, broadcastModelKey)));
+                };
+        DataStream<Row> output =
+                BroadcastUtils.withBroadcastStream(
+                        Collections.singletonList(input), broadcastMap, function);
+
+        Table outputTable = tEnv.fromDataStream(output);
+
+        return new Table[] {outputTable};
+    }
+
+    @Override
+    public void save(String path) throws IOException {
+        StreamTableEnvironment tEnv =
+                (StreamTableEnvironment) ((TableImpl) modelTable).getTableEnvironment();
+
+        String dataPath = ReadWriteUtils.getDataPath(path);
+        FileSink<NaiveBayesModelData> sink =
+                FileSink.forRowFormat(
+                                new Path(dataPath), new NaiveBayesModelData.ModelDataEncoder())
+                        .withRollingPolicy(OnCheckpointRollingPolicy.build())
+                        .withBucketAssigner(new BasePathBucketAssigner<>())
+                        .build();
+        NaiveBayesModelData.toDataStream(tEnv, modelTable).sinkTo(sink);
+
+        ReadWriteUtils.saveMetadata(this, path);
+    }
+
+    public static NaiveBayesModel load(StreamExecutionEnvironment env, String path)
+            throws IOException {
+        StreamTableEnvironment tEnv = StreamTableEnvironment.create(env);
+
+        Source<NaiveBayesModelData, ?, ?> source =
+                FileSource.forRecordStreamFormat(
+                                new NaiveBayesModelData.ModelDataStreamFormat(),
+                                ReadWriteUtils.getDataPaths(path))
+                        .build();
+        NaiveBayesModel model = ReadWriteUtils.loadStageParam(path);
+        DataStream<NaiveBayesModelData> modelData =
+                env.fromSource(source, WatermarkStrategy.noWatermarks(), "modelData");
+        model.setModelData(NaiveBayesModelData.fromDataStream(tEnv, modelData));
+
+        return model;
+    }
+
+    @Override
+    public Map<Param<?>, Object> getParamMap() {
+        return paramMap;
+    }
+
+    @Override
+    public NaiveBayesModel setModelData(Table... inputs) {
+        Preconditions.checkArgument(inputs.length == 1);
+        modelTable = inputs[0];
+        return this;
+    }
+
+    @Override
+    public Table[] getModelData() {
+        return new Table[] {modelTable};
+    }
+
+    private static class PredictLabelOperator
+            extends AbstractUdfStreamOperator<Row, PredictLabelFunction>
+            implements OneInputStreamOperator<Row, Row> {
+        public PredictLabelOperator(PredictLabelFunction userFunction) {
+            super(userFunction);
+        }
+
+        @Override
+        public void processElement(StreamRecord<Row> streamRecord) {
```

Comment:
Would it be simpler to remove `PredictLabelFunction` and move its logic into `PredictLabelOperator::processElement`?

Reply:
I'm afraid that we cannot achieve this by now. The API `withBroadcast` requires that I must pass to it a child class of `AbstractUdfStreamOperator`. A child class of `AbstractStreamOperator` would throw exception at runtime.

Reply:
Actually you can make the change if you use the RichMapFunction.
But I think making it a separate function is more clear to me.

Reply:
Actually you can make the change if you use the RichMapFunction.
But I think making it a separate function is more clear to me.

Reply:
> 

Ah, I see. 
Yes, the broadcast variables can only be accessed by `RichFunction#getRuntimeContext#getBroadcastVariable` now. 

A possible solution to put processing logic in `PredictLabelOperator::processElement` is to add a AbstractFunction but only use it for accessing broadcast variables. An examples can be found in `org.apache.flink.ml.common.broadcast.operator.TestOneInputOp`

Reply:
You are right. This method is feasible. I have make this change and removed `PredictLabelFunction`.